src/rust/catnap/linux/transport.rs

```diff
@@ -504,8 +476,151 @@ impl SharedCatnapTransport {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        self.0.deref_mut()
+    }
+}
+
+/// Turn a shared socket metadata structure into the raw file descriptor.
+impl AsRawFd for SharedSocketData {
+    fn as_raw_fd(&self) -> RawFd {
+        self.get_socket().as_raw_fd()
+    }
+}
+
+/// Clean up the epoll socket on libOS shutdown.
+impl Drop for CatnapTransport {
+    fn drop(&mut self) {
+        self.background_task
+            .wake_with(Err(Fail::new(libc::EBADF, "closing epoll socket")));
+        match unsafe { libc::close(self.epoll_fd) } {
+            0 => (),
+            -1 => {
+                let errno: libc::c_int = unsafe { *libc::__errno_location() };
+                if errno == libc::EBADF {
+                    warn!("epoll socket already closed");
+                } else {
+                    panic!("could not free epoll socket: {:?}", errno);
+                }
+            },
+            _ => {
+                unreachable!("close can only return 0 or -1");
+            },
+        }
+    }
+}
+
+/// Dereference a shared reference to socket metadata.
+impl Deref for SharedSocketData {
+    type Target = SocketData;
+
+    fn deref(&self) -> &Self::Target {
+        self.0.deref()
+    }
+}
+
+/// Dereference a shared reference to socket metadata.
+impl DerefMut for SharedSocketData {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        self.0.deref_mut()
+    }
+}
+
+/// Turn a shared reference for socket metadata into a direct reference to the metadata.
+impl AsRef<SocketData> for SharedSocketData {
+    fn as_ref(&self) -> &SocketData {
+        self.0.as_ref()
+    }
+}
+
+/// Turn a shared mutable reference for socket metadata into a direct mutable reference to the metadata.
+impl AsMut<SocketData> for SharedSocketData {
+    fn as_mut(&mut self) -> &mut SocketData {
+        self.0.as_mut()
+    }
+}
+
+impl NetworkTransport for SharedCatnapTransport {
+    type SocketDescriptor = usize;
+
+    /// Create a new Linux-based network transport.
+    fn new(_config: &Config, runtime: &mut SharedDemiRuntime) -> Self {
+        // Create epoll socket.
+        // Linux ignores the size argument to epoll, it just has to be more than 0.
+        let epoll_fd: RawFd = match unsafe { libc::epoll_create(10) } {
+            fd if fd >= 0 => fd.into(),
+            _ => {
+                let errno: libc::c_int = unsafe { *libc::__errno_location() };
+                panic!("could not create epoll socket: {:?}", errno);
+            },
+        };
+
+        // Set up background task for polling epoll API.
+        let yielder: Yielder = Yielder::new();
+        let background_task: YielderHandle = yielder.get_handle();
+        let me: Self = Self(SharedObject::new(CatnapTransport {
+            epoll_fd,
+            socket_table: Slab::<SharedSocketData>::new(),
+            background_task,
+        }));
+        let mut me2: Self = me.clone();
+        runtime
+            .insert_background_coroutine(
```

Comment:
The background coroutine is woken up with error in the drop() fn and the poll coroutine exits on any error on yielder await. Should the insert_background_coroutine() be paired with remove_background_coroutine() too? (for proper cleanup inside the scheduler)

Reply:
Sure, I couldn't remember if background coroutines were always removed once they exited. But I still want to wake the background coroutine so that it can do any clean up needed.